packages/go-gen/src/schema.rs

```diff
@@ -0,0 +1,272 @@
+                    docs: documentation(schema),
+                    ty,
+                })
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let strct = GoStruct {
+            name: new_struct_name.clone(),
+            docs: None,
+            fields,
+        };
+        additional_structs.push(strct);
+
+        new_struct_name
+    } else if t.contains(&InstanceType::Array) {
+        // get type of items
+        let item_type = array_item_type(schema, type_context, additional_structs)
+            .context("failed to get array item type")?;
+
+        // for nullable array item types, we have to use a pointer type, even for basic types,
+        // so we can pass null as elements
+        // otherwise they would just be omitted from the array
+        replace_custom_type(&if item_type.is_nullable {
+            format!("[]*{}", item_type.name)
+        } else {
+            format!("[]{}", item_type.name)
+        })
+    } else {
+        unreachable!("instance type should be one of the above")
+    })
+}
+
+/// Extract the type of the items of an array.
+///
+/// This fails if the given schema object is not an array,
+/// has multiple item types or other errors occur during type extraction of
+/// the underlying schema.
+pub fn array_item_type(
+    schema: &SchemaObject,
+    type_context: TypeContext,
+    additional_structs: &mut Vec<GoStruct>,
+) -> Result<GoType> {
+    match schema.array.as_ref().and_then(|a| a.items.as_ref()) {
+        Some(SingleOrVec::Single(array_validation)) => {
+            schema_object_type(array_validation.object()?, type_context, additional_structs)
+        }
+        _ => bail!("array type with non-singular item type is not supported"),
+    }
+}
+
+/// Tries to extract a type name from the given subschemas.
+///
+/// This fails if there are multiple subschemas or other errors occur
+/// during subschema type extraction.
+pub fn subschema_type(
+    subschemas: &[Schema],
+    type_context: TypeContext,
+    additional_structs: &mut Vec<GoStruct>,
+) -> Result<String> {
+    ensure!(
+        subschemas.len() == 1,
+        "multiple subschemas are not supported"
+    );
+    let subschema = &subschemas[0];
+    let GoType { name, .. } =
+        schema_object_type(subschema.object()?, type_context, additional_structs)?;
+    Ok(replace_custom_type(&name))
+}
+
+pub fn is_null(schema: &SchemaObject) -> bool {
+    schema
+        .instance_type
+        .as_ref()
+        .map_or(false, |s| s.contains(&InstanceType::Null))
+}
+
+pub fn documentation(schema: &SchemaObject) -> Option<String> {
+    schema.metadata.as_ref()?.description.as_ref().cloned()
+}
+
+/// Maps special types to their Go equivalents.
+/// If the given type is not a special type, returns `None`.
+pub fn custom_type_of(ty: &str) -> Option<&str> {
+    match ty {
+        "Uint128" => Some("string"),
+        "Binary" => Some("[]byte"),
+        "HexBinary" => Some("Checksum"),
```

Comment:
This should be String by default and only Checksum in cases where we have a 32 byte checksum.

Do we have a way to annotate this somehow in Rust?

Reply:
That's a bit of a problem. What we could do is use `#[schemars(title = "Checksum")]` on the field. That would set `metadata.title` for the field in the schema.

Reply:
Oh but what we also do is change `CodeInfoResponse::checksum` from `HexBinary` to `cosmwasm_std::Checksum`. With that we can map

```
// both hex strings in JSON
"HexBinary" => Some("string"),
"Checksum" => Some("Checksum"),
```

Reply:
Unfortunately, that only works if we make `Checksum` a newtype like `struct Checksum(HexBinary)`, which would be a contract-breaking change for `CodeInfoResponse`.
Type aliases are already resolved to their underlying types at the json schema level (`type Checksum = HexBinary` becomes `HexBinary` in the schema).

Reply:
Yeah, I'm thinking about a source breaking change here. In contrast to HexBinary, Checksum would guarantee to store 32 bytes. We have Checksum already and would just have to add serde.

Reply:
> we have Checksum in `cosmwasm-vm`.

Ah, right. I did not realize we don't have it in cosmwasm-std yet. But yes, this can be the same type by moving it to cosmwasm-std and adding serde support.

> wait for 2.0, right?

Yes. In the meantime we can keep `        "HexBinary" => Some("Checksum"),` because this is only used in one place.

Reply:
Alright, I created an issue for that: #1835 
Anything left to do here, or can we merge?